cmd/sourceControl/github.go

```diff
@@ -0,0 +1,158 @@
+package sourceControl
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	gh "github.com/google/go-github/github"
+	"golang.org/x/oauth2"
+	"os"
+	"strconv"
+	"strings"
+)
+
+const (
+	ghToken           = "GH_TOKEN"
+	ghRepo            = "GITHUB_REPOSITORY"
+	ghRefName         = "GITHUB_REF_NAME"
+	ghActor           = "GITHUB_ACTOR"
+	ghSha             = "GITHUB_SHA"
+	ghEvent           = "GITHUB_EVENT_NAME"
+	ghRef             = "GITHUB_REF"
+	ghTriggeringActor = "GITHUB_TRIGGERING_ACTOR"
+	ghRefType         = "GITHUB_REF_TYPE"
+	ghServer          = "GITHUB_SERVER_URL"
+	ghRunId           = "GITHUB_RUN_ID"
+	ghWorkflow        = "GITHUB_WORKFLOW"
+)
+
+type (
+	GithubContext struct {
+		BaseContext
+		Github  GithubData `json:"github,omitempty"`
+		service GithubService
+	}
+
+	GithubData struct {
+		RunId    string `json:"runId,omitempty"`
+		Workflow string `json:"workflow,omitempty"`
+	}
+
+	GithubService interface {
+		GetPR() (gh.PullRequest, error)
+	}
+	GithubClient interface {
+		getPR(owner string, repo string, number int) (*gh.PullRequest, error)
+		searchForPr(options *gh.PullRequestListOptions) (*gh.PullRequest, error)
+		init(token string)
+	}
+	DefaultGithubService struct {
+		client GithubClient
+	}
+
+	DefaultGithubClient struct {
+		client gh.Client
+		ctx    context.Context
+	}
+)
+
+func (gc GithubContext) GetContext() (Context, error) {
+	githubContext := GithubContext{
+		BaseContext: BaseContext{
+			Type:                github,
+			Event:               Event(os.Getenv(ghEvent)),
+			Reference:           Reference(os.Getenv(ghRefType)),
+			ReferenceName:       os.Getenv(ghRefName),
+			Principal:           os.Getenv(ghActor),
+			TriggeringPrincipal: os.Getenv(ghTriggeringActor),
+			Sha:                 os.Getenv(ghSha),
+			Repository:          os.Getenv(ghRepo),
+			Server:              os.Getenv(ghServer)},
+		Github: GithubData{
+			RunId:    os.Getenv(ghRunId),
+			Workflow: os.Getenv(ghWorkflow),
+		}}
```

Comment:
Just to confirm that because of

```golang
	GithubContext struct {
		BaseContext
		Github  GithubData `json:"github,omitempty"`
		service GithubService
	}
```

Does this produces JSON that looks like, where base context and github are merged into a single object?

```json
{
  ...
  "scmc": {
    "type": "github",
    ...
    "github": {
      "runId": "foo",
      ...
    }
  }
}
```

Reply:
Yes, this is an example of the output:
```json
"scm":
    {
        "type": "github",
        "event": "pull_request",
        "reference": "branch",
        "referenceName": "5/merge",
        "source": "pulling",
        "target": "main",
        "principal": "cs-simental",
        "triggeringPrincipal": "cs-simental",
        "prTitle": "test valid token",
        "prUrl": "https://github.com/cs-simental/potato-facts/pull/5",
        "sha": "ebc38b76b1eba3fcb27be7af82064e8a1c988bc",
        "repository": "cs-simental/potato-facts",
        "server": "https://github.com",
        "github":
        {
            "runId": "234568",
            "workflow": "Deploy to CD-as-a-Service"
        }
```